src/bp/core/services/dialog/decision-engine.ts

```diff
@@ -25,75 +31,102 @@ export class DecisionEngine {
+    }
+
+    const elected = event.suggestions!.find(x => x.decision.status === 'elected')
+
+    if (elected) {
+      Object.assign(event, { decision: elected })
+      await this._sendSuggestion(elected, sessionId, event)
     }
 
     if (!event.hasFlag(WellKnownFlags.SKIP_DIALOG_ENGINE)) {
       try {
+        Object.assign(event, {
+          decision: <IO.Suggestion>{
+            decision: { reason: 'no suggestion matched', status: 'elected' },
+            confidence: 1,
+            payloads: [],
+            source: 'decisionEngine',
+            sourceDetails: 'execute default flow'
+          }
+        })
         const processedEvent = await this.dialogEngine.processEvent(sessionId, event)
         await this.stateManager.persist(processedEvent, false)
       } catch (err) {
         this.logger
           .forBot(event.botId)
           .attachError(err)
-          .error('An unexpected flow error occurred.')
+          .error('An unexpected error occurred.')
         await this._sendErrorMessage(event)
       }
     }
   }
 
-  protected _findBestReply(event: IO.IncomingEvent): IO.Suggestion | undefined {
-    const replies = _.sortBy(event.suggestions, reply => -reply.confidence)
-    const lastMsg = _.last(event.state.session.lastMessages)
+  protected _amendSuggestionsWithDecision(suggestions: IO.Suggestion[], turnsHistory: IO.DialogTurnHistory[]) {
+    const replies = _.sortBy(suggestions, ['confidence'], ['DESC'])
+    const lastMsg = _.last(turnsHistory)
+    const lastMessageSource = lastMsg && lastMsg.replySource
 
-    // If the user asks the same question, chances are he didnt get the response he wanted.
-    // So we cycle through the other suggestions and return the next best reply with a high enough confidence.
-    for (let i = 0; i < replies.length; i++) {
-      const bestReplyIntent = replies[i].intent
-      const lastMessageIntent = lastMsg && lastMsg.intent
-
-      if (bestReplyIntent === lastMessageIntent) {
-        const nextBestReply = replies[i + 1]
+    let bestReply: IO.Suggestion | undefined = undefined
 
-        if (this._isConfidentReply(nextBestReply)) {
-          return nextBestReply
-        } else {
-          return // If confidence is too low, we dont need to check other replies
-        }
+    for (let i = 0; i < replies.length; i++) {
+      const replySource = replies[i].source + ' ' + replies[i].sourceDetails || Date.now()
+
+      const violatesRepeatPolicy =
+        replySource === lastMessageSource &&
+        moment(lastMsg!.replyDate)
+          .add(this.MIN_NO_REPEAT, 'ms')
+          .isAfter(moment())
+
+      if (replies[i].confidence < this.MIN_CONFIDENCE) {
+        replies[i].decision = { status: 'dropped', reason: `confidence lower than ${this.MIN_CONFIDENCE}` }
+      } else if (violatesRepeatPolicy) {
+        replies[i].decision = { status: 'dropped', reason: `bot would repeat itself (within ${this.MIN_NO_REPEAT}ms)` }
```

Comment:
In this case we should be giving "next best" if confidence is higher than min confidence but I don't see how this could be working with this method

Reply:
The loop will do it right?  Next iteration.  The trick is they are sorted by confidence so the next result (if confidence is higher than MIN) is going to be elected.

Reply:
Right ! Sorted !